### DISCUSSION TITLE
Static analysis to detect node usage before migrating to webpack 5

### DISCUSSION THREAD

**Discussion by bdwain** [webpack/webpack#13313] [Q&A]
Has anyone done any sort of static analysis to detect node usage when migrating to webpack 5? (For this https://webpack.js.org/migrate/5/#test-webpack-5-compatibility) 

I'm having trouble finding where it's used throughout my node modules, and am worried about moving forward with the migration with so little confidence because the issues don't pop up until runtime.

If there was some sort of plugin or option for webpack 4 that could alert you to when node usages were detected, that would make the transition much easier. Since webpack 4 already detects node usage when deciding whether to include the node plugin at all, it seems like it should be easy enough to hook into that.

**Comment by alexander-akait** [comment]
webpack says about node module usage, you will see warnings

**Reply by bdwain** [reply]
Is there something you need to do to enable these checks? I am not seeing them after running a webpack build, even if the app uses `process` or `Buffer` under the hood

**Reply by bdwain** [reply]
@alexander-akait I think I realized the issue. I came across this error

```
BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.
```

But that seems to only occur when importing core node libraries. `process` and `Buffer` are globals, so they don't have those errors that appear at build time. Only at runtime, which are much harder to detect.

**Reply by alexander-akait** [reply]
Yep, here a problem we can found `process`/`Buffer`, but sometimes developers do this (just simple examples):
```
if (process.env.NODE_ENV === 'production') { /* Conde */ }
const blob = isNode ? Buffer.alloc(5) : [0, 0, 0, 0];  
```

And we will have a lot of false positive warnings, just compile code with `minimize: false` and search usage `process` and `Buffer`, most of libraries have `browser`/`exports`/checks in code so it is not big problem. Also very few packages used `Buffer` or `process` without `require` Node.js modules.

**Reply by bdwain** [reply]
I see. I just would like to be able to automate this check ideally so that if someone accidentally adds a dependency that uses node, the build fails and we don't just find out at runtime after the app is deployed to production. I guess imports of node modules are already covered by webpack's internals, but globals seem to be a blind spot.

What are your thoughts on a new webpack configuration option (or plugin) that does the searching of the code for usages of node globals (`Buffer`, `process`, any others that there are) and warns (or maybe errors) if they are detected. Then it could be a build time check and give people confidence going forward.
